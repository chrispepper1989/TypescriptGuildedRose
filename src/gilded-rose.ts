export type ValidItems = 
    "+5 Dexterity Vest"
    | "Aged Brie" 
    | "Elixir of the Mongoose"
    |"Sulfuras, Hand of Ragnaros"
    | "Backstage passes to a TAFKAL80ETC concert"
    | "Conjured Mana Cake"

export class Item {
    name: ValidItems;
    sellIn: number;
    quality: number;

    constructor(name: ValidItems, sellIn: number, quality: number) {
        this.name = name;
        this.sellIn = sellIn;
        this.quality = quality;
    }
}

export class GildedRose {
    items: Array<Item>;

    constructor(items = [] as Array<Item>) {
        this.items = items;
    }
    
    updateQuality() {
        return GildedRose.updateItemsQuality(this.items);
    }

    private static updateItemsQuality(items:Item[]) {
        items.forEach(item => {
            switch (item.name) {
                case "Aged Brie":
                    this.updateAgedBrieItem(item);
                    break;
                case "+5 Dexterity Vest":
                    this.updateOtherItems(item);
                    break;
                case "Elixir of the Mongoose":                  
                case "Sulfuras, Hand of Ragnaros":             
                case "Backstage passes to a TAFKAL80ETC concert":
                case "Conjured Mana Cake":
                default:
                    this.updateOtherItems(item);
                    break;
            }
        });

        return items;
    }

    private static updateAgedBrieItem(item: Item) {
        if(item.name !== "Aged Brie")
            return;
        
        
        if (item.quality < 50) {
            item.quality = item.quality + 1            
        }
                
        item.sellIn = item.sellIn - 1;
      
        if (item.sellIn < 0) {
            if (item.quality < 50) {
                item.quality = item.quality + 1
            }
        }
    }
    
    private static updateOtherItems(item: Item) {
        if(item.name === "Aged Brie")
            return;
        if (item.name != 'Backstage passes to a TAFKAL80ETC concert') {
            if (item.quality > 0) {
                if (item.name != 'Sulfuras, Hand of Ragnaros') {
                    item.quality = item.quality - 1
                }
            }
        } else {
            if (item.quality < 50) {
                item.quality = item.quality + 1
                if (item.name == 'Backstage passes to a TAFKAL80ETC concert') {
                    if (item.sellIn < 11) {
                        if (item.quality < 50) {
                            item.quality = item.quality + 1
                        }
                    }
                    if (item.sellIn < 6) {
                        if (item.quality < 50) {
                            item.quality = item.quality + 1
                        }
                    }
                }
            }
        }
        if (item.name != 'Sulfuras, Hand of Ragnaros') {
            item.sellIn = item.sellIn - 1;
        }
        if (item.sellIn < 0) {
            if (item.name != 'Backstage passes to a TAFKAL80ETC concert') {
                if (item.quality > 0) {
                    if (item.name != 'Sulfuras, Hand of Ragnaros') {
                        item.quality = item.quality - 1
                    }
                }
            } else {
                item.quality = item.quality - item.quality
            }
        }
    }
}
