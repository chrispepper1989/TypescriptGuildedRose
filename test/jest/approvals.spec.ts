import { Item, GildedRose } from '@/gilded-rose';

/**
 * This unit test uses [Jest Snapshot](https://goo.gl/fbAQLP).
 * 
 * There are two test cases here with different styles:
 * <li>"foo" is more similar to the unit test from the 'Java' version
 * <li>"thirtyDays" is more similar to the TextTest from the 'Java' version
 *
 * I suggest choosing one style to develop and deleting the other.
 */
describe('Gilded Rose Approval (individual tests)', () => {

describe('Items', () => {
 
  /* coverage for all known items over 30 days*/
  test.each([10,30,40])
  ('when given all the known items over %p days, return items based on current rules  ', (days:number) => {

    const sentItems = [
      new Item("+5 Dexterity Vest", 10, 20), //
      new Item("Aged Brie", 2, 0), //
      new Item("Elixir of the Mongoose", 5, 7), //
      new Item("Sulfuras, Hand of Ragnaros", 0, 80), //
      new Item("Sulfuras, Hand of Ragnaros", -1, 80),
      new Item("Backstage passes to a TAFKAL80ETC concert", 15, 20),
      new Item("Backstage passes to a TAFKAL80ETC concert", 10, 49),
      new Item("Backstage passes to a TAFKAL80ETC concert", 5, 49)]
    
    const gildedRose = new GildedRose(sentItems);
    
    Array.from({ length: days }, x => gildedRose.updateQuality())
    
    const finalItems = gildedRose.updateQuality();
    expect(finalItems).toMatchSnapshot();
  });
})});

