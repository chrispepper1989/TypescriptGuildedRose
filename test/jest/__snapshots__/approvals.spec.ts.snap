// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gilded Rose Approval (individual tests) Items when given all the known items over 10 days, return items based on current rules   1`] = `
[
  Item {
    "name": "+5 Dexterity Vest",
    "quality": 8,
    "sellIn": -1,
  },
  Item {
    "name": "Aged Brie",
    "quality": 20,
    "sellIn": -9,
  },
  Item {
    "name": "Elixir of the Mongoose",
    "quality": 0,
    "sellIn": -6,
  },
  Item {
    "name": "Sulfuras, Hand of Ragnaros",
    "quality": 80,
    "sellIn": 0,
  },
  Item {
    "name": "Sulfuras, Hand of Ragnaros",
    "quality": 80,
    "sellIn": -1,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 38,
    "sellIn": 4,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 0,
    "sellIn": -1,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 0,
    "sellIn": -6,
  },
]
`;

exports[`Gilded Rose Approval (individual tests) Items when given all the known items over 30 days, return items based on current rules   1`] = `
[
  Item {
    "name": "+5 Dexterity Vest",
    "quality": 0,
    "sellIn": -21,
  },
  Item {
    "name": "Aged Brie",
    "quality": 50,
    "sellIn": -29,
  },
  Item {
    "name": "Elixir of the Mongoose",
    "quality": 0,
    "sellIn": -26,
  },
  Item {
    "name": "Sulfuras, Hand of Ragnaros",
    "quality": 80,
    "sellIn": 0,
  },
  Item {
    "name": "Sulfuras, Hand of Ragnaros",
    "quality": 80,
    "sellIn": -1,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 0,
    "sellIn": -16,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 0,
    "sellIn": -21,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 0,
    "sellIn": -26,
  },
]
`;

exports[`Gilded Rose Approval (individual tests) Items when given all the known items over 40 days, return items based on current rules   1`] = `
[
  Item {
    "name": "+5 Dexterity Vest",
    "quality": 0,
    "sellIn": -31,
  },
  Item {
    "name": "Aged Brie",
    "quality": 50,
    "sellIn": -39,
  },
  Item {
    "name": "Elixir of the Mongoose",
    "quality": 0,
    "sellIn": -36,
  },
  Item {
    "name": "Sulfuras, Hand of Ragnaros",
    "quality": 80,
    "sellIn": 0,
  },
  Item {
    "name": "Sulfuras, Hand of Ragnaros",
    "quality": 80,
    "sellIn": -1,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 0,
    "sellIn": -26,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 0,
    "sellIn": -31,
  },
  Item {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 0,
    "sellIn": -36,
  },
]
`;
